cli/core/list.go

```diff
@@ -55,11 +56,13 @@ func runListCommand(cmd *cobra.Command, args []string) {
 
 	platforms, err := core.GetPlatforms(instance.Id, listFlags.updatableOnly)
 	if err != nil {
-		formatter.PrintError(err, "Error listing platforms")
+		feedback.Errorf("Error listing platforms: %v", err)
 		os.Exit(errorcodes.ErrGeneric)
 	}
 
-	if output.JSONOrElse(platforms) {
+	if globals.OutputFormat == "json" {
+		feedback.PrintJSON(platforms)
+	} else {
```

Comment:
I'll add a comment here but it's valid for all the occurrence of this snippet of course :-)

Is it better to write something like:

```go
	if feedback.PrintJSONOrElse(platforms) {
￼		outputInstalledCores(platforms)
	}
```

or

```go
	if globals.OutputFormat == "json" {
		feedback.PrintJSON(platforms)
	} else {
		outputInstalledCores(platforms)
	}
```

This is boilerplate code that is duplicated everywhere (basically on each command that may potentially output json).
Every breaking change to the internals of the `feedback`/`globals.OutputFormat`, just to say, must be propagated on all occurrences of this snippet.

In this case, IMHO, less is more.

Reply:
Let me see if I can find a simple way to remove the boilerplate before defending it :smile:

Reply:
Here is my proposal: let's move the responsibility to evaluate the format away from the commands by introducing a ["results container"](https://github.com/arduino/arduino-cli/blob/e57ae3ca3468c09d9434a125f67861c3ec17aaa6/cli/feedback/feedback.go#L39) that's very thin but generic enough to support different formats and that would be used only when highly customized outputs are required.

PROS:
* The boilerplate is gone, cfr [this example with no if-s at all](https://github.com/arduino/arduino-cli/blob/e57ae3ca3468c09d9434a125f67861c3ec17aaa6/cli/board/details.go#L54). As a side effect, the format can be removed from the globals.
* This approach is future proof, any new format we should add would have very little to zero impact on the existing commands
* When a complex formatting logic is needed, that stays with the command that needs it
CONS:
* When the output is complex, you have to [write a struct](https://github.com/arduino/arduino-cli/blob/massi/output/cli/board/details.go#L59) to provide the custom logic
* the feedback package gets more responsibilities but somebody has to decide which output to use

Reply:
`PrintResult` is defined as:

```
func (fb *Feedback) PrintResult(res Result) {
```

but IIUC we need a function that accepts any type and do a runtime check to see if the argument is a `Result` (unless we want to always wrap the result in a `Result` but this will multiply the boilerplate by 10x), so something like this:

```
func (fb *Feedback) PrintResult(res interface{}) {
    if r, ok := res.(Result); ok {
        ...print using JSON(r.Data()) or r.String()...
    } else {
        ...print using JSON(res) or res.String()...
    }
}
```

Anyway, besides the above, the issue is another: there isn't a generic way to print complex outputs.

One thing is a simple primitive type like `int` or `string` another is a `Library` or a `Board` or a `[]PortList`, they all have their custom print with tables and verbosity that most of the time shows a subset of the available fields, sorted in some random way, so you're basically in always in the worst case (where you need to wrap inside a `Result`).

Reply:
No type assertion, you always get a `Result` that can provide a string or an interface of some sort that's supposed to be marshalled (today in JSON, tomorrow we'll see). The "boilerplate" you add by always providing a `Result` object is just a struct definition plus a method definition (you always have an equivalent of the String method).

>Anyway, besides the above, the issue is another: there isn't a generic way to print complex outputs.
I know, I tentatively drafted a solution to see if I could remove the pattern
```
	if globals.OutputFormat == "json" {
		feedback.PrintJSON(ports)
	} else {
		outputListResp(ports)
	}
```
but I'm fine with that. Instead I'm not fine with something like `JSONOrElse` which is non-obvious, has a vague name and doesn't solve the root problem.

Reply:
> but I'm fine with that. Instead I'm not fine with something like JSONOrElse which is non-obvious, has a vague name and doesn't solve the root problem.

Ok, let's keep the bigger if then, there is no point in discussing this further, we could always go back if we change our mind ;-)